guest/api/framework/interface.go

```diff
@@ -0,0 +1,178 @@
+package framework
+
+import (
+	"context"
+	"errors"
+	"strings"
+
+	protoapi "sigs.k8s.io/kube-scheduler-wasm-extension/kubernetes/proto/api"
+)
+
+// Plugin is the parent type for all the scheduling framework plugins.
+type Plugin interface {
+	Name() string
+}
+
+// FilterPlugin is an interface for Filter plugins. These plugins are called at the
+// filter extension point for filtering out hosts that cannot run a pod.
+// This concept used to be called 'predicate' in the original scheduler.
+// These plugins should return "Success", "Unschedulable" or "Error" in Status.code.
+// However, the scheduler accepts other valid codes as well.
+// Anything other than "Success" will lead to exclusion of the given host from
+// running the pod.
+type FilterPlugin interface {
+	Plugin
+	// Filter is called by the scheduling framework.
+	// All FilterPlugins should return "Success" to declare that
+	// the given node fits the pod. If Filter doesn't return "Success",
+	// it will return "Unschedulable", "UnschedulableAndUnresolvable" or "Error".
+	// For the node being evaluated, Filter plugins should look at the passed
+	// nodeInfo reference for this particular node's information (e.g., pods
+	// considered to be running on the node) instead of looking it up in the
+	// NodeInfoSnapshot because we don't guarantee that they will be the same.
+	// For example, during preemption, we may pass a copy of the original
+	// nodeInfo object that has some pods removed from it to evaluate the
+	// possibility of preempting them to schedule the target pod.
+	Filter(ctx context.Context, state *CycleState, pod PodFn, nodeInfo NodeInfoFn) *Status
```

Comment:
I don't think we should add ctx parameter here as it won't be what people think it is. The ctx will have no value in tinygo as it isn't related to the host context.

Reply:
Well. I'd like to keep it to make the interface as similar to the scheduler's one as possible even if we cannot pass the context from host to guest. But, as you said, not to make people confused, we need to note somewhere about that at least though.

Also, making it similar to the scheduler's original interface is not only the reason. This PR [just passes](https://github.com/kubernetes-sigs/kube-scheduler-wasm-extension/blob/1da62a0ca248d800be114ef343b0576be9e98a69/guest/filter.go#L23) the plain `context.Background()` to user-defined Filter, but in the future, we may want to do something further like:
- have logger in the ctx so that people can use the logger from ctx with [FromContext](https://github.com/sanposhiho/kubernetes/blob/8badc6f8fe45e2133ca23ba00ecba94dbe04c8d8/vendor/k8s.io/klog/v2/contextual.go#L156)
  - In Kubernetes, it's very common to create logger from ctx.
- `calcel()` passed context somewhere when the guest instance is going to be deleted.

Those things could be reasons we pass `ctx` to users, what do you think?

Reply:
So this is in the guest code, and it cannot inherit context from the host. So, let's contain this discussion to the sandbox.

Since the guest can't inherit the context, the only thing that would be passed here is a global variable. Since no callback can return a modified context, that variable would never change. So, a context parameter (which has cost to have in the parameter list) will be no different than a field named context in practice.

Reply:
It is helpful to remember that we can't do anything you can't do in another language like C. There's no special thing about the host written in Go, that allows the guest to do anything go-specific. Not sure if this thinking helps, but maybe!

Reply:
Yes, I know we cannot pass context from the host. But, I mean we can pass something (like listed in [it](https://github.com/kubernetes-sigs/kube-scheduler-wasm-extension/pull/19#discussion_r1209353906)) to the user-defined `Filter` function [here](https://github.com/kubernetes-sigs/kube-scheduler-wasm-extension/blob/1da62a0ca248d800be114ef343b0576be9e98a69/guest/filter.go#L23).
It includes the design choice as well, @kerthcet what is your take on this?

Reply:
I don't think we should encourage this sort of thinking inside the guest. using context as a substitute for a global logger, or doing things like guest-specific deadlines are going to add overhead and ease people's misunderstanding of Wasm-specific concerns especially horrible performance vs native, no concurrency etc.

Reply:
here's the only example I can find of context being used in tinygo. Remember this is single threaded. Basically, main wires up this function and creates the root context which is a cancel context. There's no way to import an incoming context from the host.

https://github.com/fastly/compute-sdk-go/blob/main/fsthttp/handle.go#L17-L40

Reply:
Wait, allocating something on the guest vs allocating the same thing on the host; is there a performance difference between them? If so, that'll completely change my mind.

Reply:
Also, can you elaborate on "no concurrency"? Cannot user use goroutine in guest?

Reply:
yes there is a huge difference, this is why I raised the PR about benchmarks showing that guest take 10x as long as the host on the same proto. garbage collection is inefficient and there's only one thread. Even if we did add context as a parameter we would really want to test that it can do what we think it can.

The main thing is you can't "lift and shift" code without thinking into webassembly today for reasons like this. I would prefer to make small changes, and evaluate them with tests. Many loggers for example aren't really usable in wasm, not practically anyway. We have to unfortunately be very deliberate if we want the users' code to succeed.

My 2p.

Reply:
so again I'm ok with doing a big change, but I think we will learn more from smaller ones.

For example, it is one change to add different hooks, and another to have those hooks return pointers to values, and another to add context support (actually wired up not just mocked). If we do these things separately, and seems we have time to, we can learn one by one impact, and more importantly it is easier to write tests.

Reply:
> Also, can you elaborate on "no concurrency"? Cannot user use goroutine in guest?

So you can launch goroutines, but all execution happens on the calling thread. This means things like GC are a part of execution time https://wazero.io/languages/tinygo/#concurrency

Once the benchmarks PR merges, I'll add profiling so you can see what I mean with our example code.

Reply:
🤯 OK, that's my missed ones.

Then, I'd like to abandon this PR for now. Not saying we don't aim to make it similar to the original one, but on the road toward there, we should pile up smaller changes one by one while checking the performance.

Reply:
I think having a goal to be like the current SDK is shared, how we get there, and how much we can practically speaking are the differences. What I like about this PR is it shows how it might look. Thanks for listening!